folding-schemes/src/folding/nova/zk.rs

```diff
@@ -0,0 +1,500 @@
+        //    (U_{cf, j}, W_{cf, j})
+        // a. Compute T
+        let (cf_T, cf_cmT) = NIFS::<C2, CS2, true>::compute_cmT(
+            &nova.cf_cs_pp,
+            &nova.cf_r1cs,
+            &nova.cf_W_i,
+            &nova.cf_U_i,
+            &cf_W_j,
+            &cf_U_j,
+        )?;
+
+        // b. Compute folding challenge
+        let cf_r = RandomizedIVCProof::<C1, C2>::get_cyclefold_folding_challenge::<GC2>(
+            &mut challenges_sponge,
+            nova.pp_hash,
+            nova.cf_U_i.clone(),
+            cf_U_j.clone(),
+            cf_cmT,
+        )?;
+
+        // c. Compute fold
+        let (cf_W_r, _) = NIFS::<C2, CS2, true>::fold_instances(
+            cf_r,
+            &nova.cf_W_i,
+            &nova.cf_U_i,
+            &cf_W_j,
+            &cf_U_j,
+            &cf_T,
+            cf_cmT,
+        )?;
+
+        // d. Store folding proof
+        let cf_pi = FoldingProof { cmT: cf_cmT };
+
+        Ok(RandomizedIVCProof {
+            U_i: nova.U_i.clone(),
+            u_i: nova.u_i.clone(),
+            U_r,
+            pi,
+            pi_prime,
+            W_i_prime,
+            cf_U_i: nova.cf_U_i.clone(),
+            cf_pi,
+            cf_U_j,
+            cf_W_r,
+        })
+    }
+
+    #[allow(clippy::too_many_arguments)]
+    pub fn verify<
+        CS1: CommitmentScheme<C1, true>,
+        GC2: CurveVar<C2, CF2<C2>>,
+        CS2: CommitmentScheme<C2, true>,
+    >(
+        r1cs: R1CS<C1::ScalarField>,
+        cf_r1cs: R1CS<C2::ScalarField>,
+        pp_hash: C1::ScalarField,
+        poseidon_config: PoseidonConfig<C1::ScalarField>,
+        i: C1::ScalarField,
+        z_0: Vec<C1::ScalarField>,
+        z_i: Vec<C1::ScalarField>,
+        proof: &RandomizedIVCProof<C1, C2>,
+    ) -> Result<(), Error>
+    where
+        <C1 as Group>::ScalarField: Absorb,
+        <C2 as Group>::ScalarField: Absorb,
+        <C2 as CurveGroup>::BaseField: PrimeField,
+        <C2 as CurveGroup>::BaseField: Absorb,
+        for<'a> &'a GC2: GroupOpsBounds<'a, C2, GC2>,
+        GC2: ToConstraintFieldGadget<<C2 as CurveGroup>::BaseField>,
+        C1: CurveGroup<BaseField = C2::ScalarField, ScalarField = C2::BaseField>,
+    {
+        // Handles case where i=0
+        if i == C1::ScalarField::zero() {
+            if z_0 == z_i {
+                return Ok(());
+            } else {
+                return Err(Error::zkIVCVerificationFail);
+            }
+        }
+
+        // 1. Check that u_i.x is correct
+        let sponge = PoseidonSponge::<C1::ScalarField>::new(&poseidon_config);
+        let expected_u_i_x = proof.U_i.hash(&sponge, pp_hash, i, z_0, z_i);
+        if expected_u_i_x != proof.u_i.x[0] {
+            return Err(Error::zkIVCVerificationFail);
+        }
+
+        // 2. Check that u_i values are correct
+        if !proof.u_i.cmE.is_zero() || proof.u_i.u != C1::ScalarField::one() {
+            return Err(Error::zkIVCVerificationFail);
+        }
```

Comment:
`u_i` is used to compute the folding challenge `r` and `U_f`. If the verifier does not perform this check, the prover could "choose" whatever value fits him for `u_i.cmE` and `u_i.u`, manipulating the challenge `r` and the subsequent `U_f` value. Also, from a code perspective, I'm trying to get the code to stick as much as possible to the paper!

Reply:
hmmm unsure.. Since you are just checking that these aren't 0 or 1. But nothing else. 
And if they were 0 or 1, I assume somewhere else something wouldn't match. I don't recall seeing in the paper a check exactly like this one.

Hence, my question: Is this a safety-check to ensure correctness? Or is it really a requirement to verify the proof correctly?

Reply:
This is a requirement to verify the proof, the second one given p. 56